### PR TITLE
2480 partition aware subscription

### PR DESCRIPTION
**Pull request by longma1**
Added multi-tenancy support for subscriptions

### CODE REVIEW

**Review by katiesmilecdr — APPROVED** [review]

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/submit/interceptor/SubscriptionMatcherInterceptor.java [hunk 1]
```diff
@@ -114,7 +118,8 @@ public void resourceUpdated(IBaseResource theOldResource, IBaseResource theNewRe
 	 */
 	@Override
 	public void submitResourceModified(IBaseResource theNewResource, ResourceModifiedMessage.OperationTypeEnum theOperationType, RequestDetails theRequest) {
-		ResourceModifiedMessage msg = new ResourceModifiedMessage(myFhirContext, theNewResource, theOperationType, theRequest);
+		RequestPartitionId requestPartitionId = myRequestPartitionHelperSvc.determineReadPartitionForRequestForRead(theRequest, theNewResource.getIdElement().getResourceType(), theNewResource.getIdElement());
```
Comment (fil512) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/submit/interceptor/SubscriptionMatcherInterceptor.java:121]:
```suggestion
// Even though the resource is being written, the subscription will be interacting with it by effectively "reading" it so we set the RequestPartitionId as a read request
		RequestPartitionId requestPartitionId = myRequestPartitionHelperSvc.determineReadPartitionForRequestForRead(theRequest, theNewResource.getIdElement().getResourceType(), theNewResource.getIdElement());
```

File: hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/match/deliver/BaseSubscriptionDeliverySubscriberTest.java [hunk 2]
```diff
@@ -181,4 +176,77 @@ public void testInterceptorBroadcasterAbortsDelivery() {
...
+		assertThat(jsonString, containsString("\"canonicalSubscription\":"));
+
+		// Assert that the default partitionID is being generated and is being serialized in JSON
+		assertThat(jsonString, containsString("\"allPartitions\":false"));
+		assertThat(jsonString, containsString("\"partitionIds\":[null]"));
+	}
+
+	@Test
+	public void testSerializeLegacyDeliveryMessage() throws JsonProcessingException {
+		String legacyDeliveryMessageJson = "{\"headers\":{\"retryCount\":0,\"customHeaders\":{}},\"payload\":{\"operationType\":\"CREATE\",\"canonicalSubscription\":{\"id\":\"Subscription/123\",\"endpointUrl\":\"http://example.com/fhir\",\"payload\":\"application/fhir+json\"},\"payload\":\"{\\\"resourceType\\\":\\\"Patient\\\",\\\"active\\\":true}\"}}";
```
Comment (fil512) [inline hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/match/deliver/BaseSubscriptionDeliverySubscriberTest.java:227]: Fine to leave this as is.  But generally for stuff like this, it's better to put the json out in a test/resources/legacy-subscription.json file

File: hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/subscriber/SubscriptionMatchingSubscriberTest.java [hunk 3]
```diff
@@ -121,5 +148,127 @@ public void testCriteriaStarOnly() throws InterruptedException {
 		assertEquals(Constants.CT_FHIR_XML_NEW, ourContentTypes.get(0));
 	}
 
+	@Test
```
Comment (fil512) [inline hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/subscriber/SubscriptionMatchingSubscriberTest.java:151]: Excellent tests!

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/partition/AllPartitionSystemRequestDetails.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import ca.uhn.fhir.interceptor.model.RequestPartitionId;
+
+public class AllPartitionSystemRequestDetails extends SystemRequestDetails{
+	public AllPartitionSystemRequestDetails(){
```
Comment (fil512) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/partition/AllPartitionSystemRequestDetails.java:26]: remove this class and create a static method:

`public static SystemRequestDetails SystemRequestDetails.forAllPartitions()`

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/subscription/BaseSubscriptionsR4Test.java [hunk 5]
```diff
@@ -143,19 +143,24 @@ public void initializeOurCountHolder() {
 
 
 	protected Observation sendObservation(String code, String system) {
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/subscription/BaseSubscriptionsR4Test.java:145]: Maybe fixup these param names while you're here.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java [hunk 6]
```diff
@@ -124,6 +124,10 @@ private CanonicalSubscription canonicalizeDstu3(IBaseResource theSubscription) {
 			if (status != null) {
 				retVal.setStatus(org.hl7.fhir.r4.model.Subscription.SubscriptionStatus.fromCode(status.toCode()));
 			}
+			RequestPartitionId requestPartitionId = (RequestPartitionId) theSubscription.getUserData(Constants.RESOURCE_PARTITION_ID);
```
Comment (michaelabuckley) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java:127]: These lines can probably be shared between R4, R5, etc.  `getUserData()` is on `IAnyResource` which is a shared base.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/CanonicalSubscription.java [hunk 7]
```diff
@@ -160,7 +162,7 @@ public void setHeaders(String theHeaders) {
 	public String getChannelExtension(String theUrl) {
 		String retVal = null;
 		List<String> strings = myChannelExtensions.get(theUrl);
-		if (strings != null && strings.isEmpty() == false) {
+		if (strings != null && !strings.isEmpty()) {
```
Comment (michaelabuckley) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/CanonicalSubscription.java:165]: `CollectionUtils.isEmpty()` may be easier to read here.

**Review by michaelabuckley — APPROVED** [review]
Looks good!

File: hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/subscriber/SubscriptionMatchingSubscriberTest.java (hunk 3 above)
Comment (StevenXLi) [reply hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/subscriber/SubscriptionMatchingSubscriberTest.java:151]: Thank you!

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/CanonicalSubscription.java (hunk 7 above)
Comment (StevenXLi) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/CanonicalSubscription.java]: changed the name of the list for better clarity

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java (hunk 6 above)
Comment (StevenXLi) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java]: extracted out the code for setting the `RequestPartitionId` to a method

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/subscription/BaseSubscriptionsR4Test.java (hunk 5 above)
Comment (StevenXLi) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/subscription/BaseSubscriptionsR4Test.java]: fixed!

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/submit/interceptor/SubscriptionMatcherInterceptor.java (hunk 1 above)
Comment (StevenXLi) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/submit/interceptor/SubscriptionMatcherInterceptor.java:122]: added the comment

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/package-info.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Module to support Subscriptions
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/package-info.java:2]: Thanks!

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionMatchingSubscriber.java [hunk 9]
```diff
@@ -123,7 +123,12 @@ private void doMatchActiveSubscriptionsAndDeliver(ResourceModifiedMessage theMsg
 		boolean resourceMatched = false;
 
 		for (ActiveSubscription nextActiveSubscription : subscriptions) {
-
+			// skip if the partitions don't match
+			CanonicalSubscription subscription = nextActiveSubscription.getSubscription();
+			if (subscription != null && subscription.getPartitionId() != null && theMsg.getPartitionId() != null
+				&& !theMsg.getPartitionId().hasPartitionId(subscription.getPartitionId())) {
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionMatchingSubscriber.java]: Yep, it's a RequestPartitionId, it's set in `SubscriptionCanonicalizer`. Fixed in the latest commit

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/matching/DaoSubscriptionMatcher.java [hunk 10]
```diff
@@ -66,15 +69,17 @@ public InMemoryMatchResult match(CanonicalSubscription theSubscription, Resource
...
 		IFhirResourceDao<?> subscriptionDao = myDaoRegistry.getSubscriptionDao();
 		RuntimeResourceDefinition responseResourceDef = subscriptionDao.validateCriteriaAndReturnResourceDefinition(theCriteria);
 		SearchParameterMap responseCriteriaUrl = myMatchUrlService.translateMatchUrl(theCriteria, responseResourceDef);
 
 		IFhirResourceDao<? extends IBaseResource> responseDao = myDaoRegistry.getResourceDao(responseResourceDef.getImplementingClass());
 		responseCriteriaUrl.setLoadSynchronousUpTo(1);
 
-		return responseDao.search(responseCriteriaUrl);
+		PartitionablePartitionId partitionId = new PartitionablePartitionId(theSubscription.getPartitionId(), null);
+		RequestDetails systemRequestDetails = new SystemRequestDetails().setRequestPartitionId(partitionId.toPartitionId());
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/matching/DaoSubscriptionMatcher.java:81]: Yes!

**Pull request merged**